alacritty/src/display/mod.rs

```diff
@@ -1534,6 +1560,63 @@ impl<T> DerefMut for Replaceable<T> {
     }
 }
 
+/// The frame timer state.
+pub struct FrameTimerInfo {
+    /// The optimal delay for the renderer.
+    swap_timeout: Duration,
+
+    /// The previous error that we've used to offset the `swap_delay`.
+    previous_error: i64,
+
+    /// The timestamp of the previous frame.
+    frame_timestamp: Instant,
+
+    /// The refresh rate we've used to compute the `swap_delay`.
+    refresh_interval: Duration,
+}
+
+impl FrameTimerInfo {
+    pub fn new() -> Self {
+        Self {
+            swap_timeout: Duration::ZERO,
+            previous_error: 0,
+            frame_timestamp: Instant::now(),
+            refresh_interval: Duration::ZERO,
+        }
+    }
+
+    /// Compute the delay that we should use to achieve the target frame
+    /// rate.
+    pub fn compute_timeout(&mut self, refresh_interval: Duration) -> Duration {
+        // Handle refresh rate change.
+        if self.refresh_interval != refresh_interval {
+            self.refresh_interval = refresh_interval;
+            self.frame_timestamp = Instant::now();
+            self.swap_timeout = refresh_interval;
+            return refresh_interval;
+        }
+
+        // Compute the error.
+        let error = self.refresh_interval.as_micros() as i64
+            - self.frame_timestamp.elapsed().as_micros() as i64;
+
+        // Don't update error and swap timeout if we've waited for too long.
+        if error < self.refresh_interval.as_micros() as i64 {
```

Comment:
I'm somewhat concerned about this logic here. It will cause an immediate scheduling which I guess is fine but that will mean our "clock" gets reset. Thus the new frame is not in line with our previous frames.

Shouldn't we aim to render at the next "modulo frame interval" time, so that our rendering stays on track and keeps its consistent clock?

Otherwise I'm afraid that if we regularly skip drawing (like every other frame this happens), we'd get very inconsistent latencies (not technically worse, just inconsistent).

Reply:
But I'm not using `zero`? I'm just using whatever we previously accumulated, just don't recompute and update error?

Reply:
Yeah it's fair that you're not using zero, but you're also not trying to adjust for the clock cycle at all. It doesn't really matter if you're using zero, last `swap_interval` or `swap_timeout` here, none of these will reliably match up with your previous clock cycles.